ledger.go

```diff
@@ -380,56 +383,186 @@ func (l *Ledger) PullTransactions() {
 
-		// Build list of transaction IDs
-		req := &TransactionPullRequest{
-			TransactionIds: make([][]byte, 0, l.transactions.Len()),
+		buf := bytes.NewBuffer(nil)
+		if _, err := l.transactionsSyncIndex.WriteTo(buf); err != nil {
+			logger.Error().Err(err).Msg("failed to marshal bloom filter")
+			continue
 		}
 
-		l.transactions.Iterate(func(tx *Transaction) bool {
-			req.TransactionIds = append(req.TransactionIds, tx.ID[:])
-			return true
-		})
+		var wg sync.WaitGroup
 
-		workerChan := make(chan *grpc.ClientConn, 16)
-		var workerWG sync.WaitGroup
-		workerWG.Add(cap(workerChan))
-		responseChan := make(chan *TransactionPullResponse, snowballK)
-		for i := 0; i < cap(workerChan); i++ {
-			go func() {
-				for conn := range workerChan {
-					client := NewWaveletClient(conn)
+		bfReq := &TransactionsSyncRequest{
+			Data: &TransactionsSyncRequest_Filter{
+				Filter: buf.Bytes(),
+			},
+		}
 
-					ctx, cancel := context.WithTimeout(context.Background(), conf.GetDownloadTxTimeout())
-					batch, err := client.PullTransactions(ctx, req)
-					if err != nil {
-						logger.Error().Err(err).Msg("failed to download missing transactions")
-						cancel()
+		var (
+			chunkSize uint64 = 1000
+			maxSize   uint64 = 1 << 20
+		)
+
+		for _, p := range peers {
+			wg.Add(1)
+			go func(conn *grpc.ClientConn) {
+				defer wg.Done()
+
+				ctx, cancel := context.WithTimeout(context.Background(), conf.GetDownloadTxTimeout())
+				defer cancel()
+
+				stream, err := NewWaveletClient(conn).SyncTransactions(ctx)
+				if err != nil {
+					logger.Error().Err(err).Msg("failed to create sync transactions stream")
+					return
+				}
+
+				defer func() {
+					if err := stream.CloseSend(); err != nil {
+						logger.Error().Err(err).Msg("failed to send sync transactions bloom filter")
+					}
+				}()
+
+				if err := stream.Send(bfReq); err != nil {
+					logger.Error().Err(err).Msg("failed to send sync transactions bloom filter")
+					return
+				}
 
-						responseChan <- nil
+				res, err := stream.Recv()
+				if err != nil {
+					logger.Error().Err(err).Msg("failed to receive sync transactions header")
+					return
+				}
+
+				count := res.GetTransactionsNum()
+				if count == 0 || count > maxSize {
+					return
+				}
 
+				if count > 0 {
+					logger.Info().
+						Uint64("count", count).
+						Msg("Requesting transaction(s) to sync.")
+				}
+
+				for count > 0 {
+					req := TransactionsSyncRequest_ChunkSize{
+						ChunkSize: chunkSize,
+					}
+
+					if count < chunkSize {
+						req.ChunkSize = count
+					}
+
+					if err := stream.Send(&TransactionsSyncRequest{Data: &req}); err != nil {
+						logger.Error().Err(err).Msg("failed to receive sync transactions header")
 						return
 					}
-					cancel()
 
-					responseChan <- batch
+					res, err := stream.Recv()
+					if err != nil {
+						logger.Error().Err(err).Msg("failed to receive sync transactions header")
+						return
+					}
+
+					transactions := res.GetTransactions()
```

Comment:
What do you think about calling this "`rawTransactions`" or "`transactionsRaw`" or "`txsRaw`" and the "`txs`" below as "`transactions`" to make it a bit more unambiguous ?